### ISSUE TITLE
Reset form section

### ISSUE THREAD

**Issue by slohmes** [erikras/redux-form#2239]
How do you reset the values for a redux-form FormSection, without resetting the entire form? The change action doesn't help because the fields still exist, just changed to a different value. How do I get specified fields to disappear like the reset action does for the entire form?

**Comment by gilmillasseau** [comment]
@slohmes do you solve this problem?

**Comment by MilllerTime** [comment]
I've struggled with this as well and haven't come up with a great solution.

It would be great if we could use `initialize()`, passing just a subset of field names and not re-initializing the entire form.

**Comment by gustavohenke** [comment]
@slohmes:
> How do I get specified fields to disappear like the reset action does for the entire form?

You could try using the `unregisterField` action.
From [the docs](http://redux-form.com/6.3.1/docs/api/ActionCreators.md/):

> #### `unregisterField(form:String, name:String)`
> Unregisters a field with the form.

---

@MilllerTime:
> It would be great if we could use `initialize()`, passing just a subset of field names and not re-initializing the entire form.

You *can* do this, with the `initialize` action itself.
From [the docs](http://redux-form.com/6.3.1/docs/api/ActionCreators.md/):

> #### initialize(form:String, data:Object, keepDirty:boolean)
> [...] If the `keepDirty` parameter is `true`, the values of currently dirty fields will be retained to avoid overwriting user edits.

**Comment by MilllerTime** [comment]
Good ideas @gustavohenke! A couple of questions:

> You could try using the `unregisterField` action.

In order for the user to continue working with the field, would you have to use the `registerField` action immediately after? Not completely ideal dispatching multiple sync actions, but sounds like that could work!

> > It would be great if we could use `initialize()`, passing just a subset of field names and not re-initializing the entire form.

> You can do this, with the `initialize` action itself.

I'm aware of the `keepDirty` param, but my understanding is that `initialize` simply won't touch dirty fields when that is set to true. It sounds like @slohmes wants to reset user input for a single field (dirty or not), and `keepDirty` wouldn't work for that. I described a use case where you could pass specific fields to `initialize` and it would operate on those fields only, dirty or not. Fields that weren't specified wouldn't be processed. I'm not aware of a way to achieve this via `initialize`, but please correct me if I'm missing something.

Seems like the first approach of re-registering a field might be the way to go.

**Comment by gustavohenke** [comment]
Reading this again about a month later, I don't know what I had in mind :)
Perhaps I'm just more versed in Redux Form!

Adding a new action called `resetSection` is probably the way to go. Will mark as enhancement, but for anyone who wants to tackle this issue, feel free to create a PR!

**Comment by MilllerTime** [comment]
Haha. Thinking on this a bit more as well, how about extending the current `reset` action to a signature more inline with the `touch`/`untouch` actions - allowing it to optionally work on just a subset of fields? I imagine it would look like the following, where only the first argument is required:

```JS
reset(form:String, ...fields:String)
```

Should be a non-breaking change, as just passing a form name would still reset the entire form. It would just offer more control over what is reset when needed. If a field name is given that has child fields, it would reset all deep fields under the given key. This would allow easily resetting a form section, as well as just a single field or two.

**Comment by gustavohenke** [comment]
Seems good to me 👍 

cc @erikras

**Comment by matmiz** [comment]
@gustavohenke @MilllerTime - I had the same issue, and the form's change function did the job. 
So, in my case, I have a fieldArray named 'sections' and each formSection has sectionData. 
When I wanted to reset a specific field, I took the initial data form the initialValues, and called "change" with the section path and the initial data.
`change('sections[1].setionData', intialData)`

Worked like a charm

**Comment by MilllerTime** [comment]
@matmiz `change` only works on a single field, [at least according to the docs](http://redux-form.com/6.5.0/docs/api/Props.md/#-change-field-string-value-any-function-). If that's all you need, then your solution works well, and I've done the same thing in the past. However, `reset` is nice in that it automatically looks up the initial values, without you having to type your field path twice (once for the fieldname, and again to get the corresponding initial value).

If you have multiple fields to reset,  but don't want to reset the whole form, you're stuck calling `change()` repeatedly, which is also a nuisance plus dispatching many actions at once has performance implications.

An enhanced `reset` method as described in the previous comment would be easy to use and performant for all use cases.

**Comment by matmiz** [comment]
@MilllerTime , that's not the case in `formSection`. I have some fields in this section, and when I pass the data object to the change function , it only resets the fields in this section, so I guess that's fine in my case. 
The current `reset` function won't work for me for two reasons:
1. The initial values aren't necessarily the form's initial values
2. If I changed other fields in my form, it will reset them 

I guess your suggested `reset`might be good, but I'm pleased with my solution anyway :)

**Comment by MilllerTime** [comment]
@matmiz your solution is definitely helpful for the specific issue described by @slohmes 😊

It's just not ideal when working with multiple fields which are not necessarily under a single `FormSection`. Unfortunately I've run into that scenario more than once haha.

**Comment by justin-lavelle** [comment]
This seems to work in 6.6.3, being able to change the values in a formSection, the docs should be updated to show this.
```
this.props.dispatch(change('formName', 'formSectionName', null)); 
this.props.dispatch(change('formName', 'formSectionName', {}));
this.props.dispatch(change('formName', 'formSectionName', { prop1: 'value1'}));
this.props.dispatch(change('formName', 'formSectionName.prop2', 'value2')); // this changes sets prop2 only, leaving other properties alone.
```
Now if I only touch/untouch worked with form sections...

**Comment by betoharres** [comment]
+1 just found myself needing this :)

**Comment by betoharres** [comment]
Not completely related to the issue, but I'm having a problem like that:

### How to update FormSection name prop
http://stackoverflow.com/questions/43971083/redux-form-how-to-update-formsection-name-prop

I'm trying to update `FormSection` `name` prop after I press my `New Post` button because the prop name of my FormSection holds the `postUid`, and when I press the button I generate a new `uid`, hence a new post.

Is there a **nice** way to do this?

The problem is that **even if I destroy the form state, and even if I re-register the fields, the loaded component will still use the old state.**

That's the real issue here, I don't want to refresh the page, ou unmount and mount the component. Any help? Thanks <3

**Comment by alyahmedaly** [comment]
is it possible to use reset('formName','SectionName')

**Comment by ljones87** [comment]
@justin-lavelle thank you, this worked for me after much frustration in trying out FieldArray and fields.push with my formSection (which didn't work). I used map to help me out:
 
```
  const rProps=['mainQuestion', 'weight', 'jsType', 'extraQuestions'];
    rProps.map(prop => (
      this.props.change(`question${qNumber}.${prop}`, ''),
      this.props.untouch(`question${qNumber}.${prop}`)
    ));

```
while no more efficient, at least cleaned it up a bit...of course would depend on how your formSections are set up...

**Comment by MilllerTime** [comment]
@alyahmedaly [per the latest docs](https://redux-form.com/7.2.0/docs/api/actioncreators.md/#-code-reset-form-string-code-), it doesn't appear you can do this yet 😕

@ljones87 I have a working `FieldArray` that renders fairly complex form items including a nested `FormSection`. Calling `props.fields.push({})` adds a new item for me, complete with a new working `FormSection`. Using `fields.push` should work, so you may want to investigate what else is going on to cause the issue you described. The fix you provided could be expensive - if you have a lot of items in the `FieldArray` you'll be dispatching *a lot* of redux actions which isn't great for performance. Also, just a small tip: When you want to iterate over an array of items, [Array.prototype.forEach()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) is the ideal method.  [Array.prototype.map()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) is used to transform items in an array without mutation, and creates a whole new array that you don't need. You can just replace `.map` with `.forEach` 🙂

**Comment by fantua** [comment]
Any updates on `resetSection` action?

Currently I'm cleaning section in this way:
```js
// List of fields to clean after unmount:
const Fields = ['streetLine1', 'streetLine2', 'streetLine3', 'city', 'country', 'state', 'zip'];

const mapDispatchToProps = (dispatch, { formName, sectionName }) => bindActionCreators({
    clear: () => clearFields(formName, false, false, ...Fields.map(field => `${sectionName}.${field}`))
}, dispatch);
```

Is there are any selector that allows to get list of section fields (to remove `Fields` constant)?

**Comment by erikras** [comment]
Published fix in [`v7.3.0`](https://github.com/erikras/redux-form/releases/tag/v7.3.0).
